### ISSUE TITLE
Field cannot be changed: CIDR

### ISSUE THREAD

**Issue by doubaokun** [kubernetes/kops#2830]
kops version
Version 1.6.2

When create cluster in an pre-defined subnet 10.10.32.0/20 inside an existing VPC, received the error: Field cannot be changed: CIDR.

`kops create cluster stage.cluster --state s3://cluster-status --network-cidr 10.10.32.0/20`

```
I0629 12:53:33.398443   98978 create_cluster.go:841] Using SSH public key: private/id_rsa.pub
I0629 12:53:34.238304   98978 subnets.go:183] Assigned CIDR 10.10.34.0/23 to subnet eu-west-2a
I0629 12:53:34.238328   98978 subnets.go:197] Assigned CIDR 10.10.32.0/26 to subnet utility-eu-west-2a
Previewing changes that will be made:

I0629 12:53:37.312653   98978 dns.go:91] Private DNS: skipping DNS validation
I0629 12:53:37.317943   98978 executor.go:91] Tasks: 0 done / 75 total; 34 can run
W0629 12:53:38.167222   98978 executor.go:109] error running task "VPC/stage.cluster" (9m59s remaining to succeed): Field cannot be changed: CIDR
I0629 12:53:38.167260   98978 executor.go:91] Tasks: 33 done / 75 total; 5 can run
W0629 12:53:38.446064   98978 executor.go:109] error running task "VPC/stage.cluster" (9m58s remaining to succeed): Field cannot be changed: CIDR
I0629 12:53:38.446092   98978 executor.go:91] Tasks: 37 done / 75 total; 1 can run
W0629 12:53:38.691914   98978 executor.go:109] error running task "VPC/stage.cluster" (9m58s remaining to succeed): Field cannot be changed: CIDR
I0629 12:53:38.691942   98978 executor.go:124] No progress made, sleeping before retrying 1 failed task(s)
```

```
I0629 12:59:56.396055   99187 executor.go:91] Tasks: 37 done / 75 total; 1 can run
I0629 12:59:56.396107   99187 executor.go:157] Executing task "VPC/stage.cluster": *awstasks.VPC {"Name":"stage.cluster","ID":"vpc-c80b8ba1","CIDR":"10.10.32.0/20","EnableDNSHostnames":null,"EnableDNSSupport":true,"Shared":true,"Tags":{"KubernetesCluster":"stage.cluster","Name":"stage.cluster","kubernetes.io/cluster/stage.cluster":"shared"}}
I0629 12:59:56.396304   99187 request_logger.go:45] AWS request: ec2/DescribeVpcs
I0629 12:59:56.586846   99187 vpc.go:82] found matching VPC *awstasks.VPC {"Name":"aws-vpc","ID":"vpc-c80b8ba1","CIDR":"10.10.0.0/16","EnableDNSHostnames":null,"EnableDNSSupport":null,"Shared":null,"Tags":{"Name":"aws-vpc"}}
I0629 12:59:56.586951   99187 request_logger.go:45] AWS request: ec2/DescribeVpcAttribute
I0629 12:59:56.688154   99187 request_logger.go:45] AWS request: ec2/DescribeVpcAttribute
I0629 12:59:56.769476   99187 changes.go:80] Field changed "Name" actual="aws-vpc" expected="stage.cluster"
I0629 12:59:56.769503   99187 changes.go:80] Field changed "CIDR" actual="10.10.0.0/16" expected="10.10.32.0/20"
I0629 12:59:56.769525   99187 changes.go:80] Field changed "Tags" actual="map[Name:aws-vpc]" expected="map[kubernetes.io/cluster/stage.cluster:shared KubernetesCluster:stage.cluster Name:stage.cluster]"
W0629 12:59:56.769590   99187 executor.go:109] error running task "VPC/stage.cluster" (9m38s remaining to succeed): Field cannot be changed: CIDR
I0629 12:59:56.769606   99187 executor.go:124] No progress made, sleeping before retrying 1 failed task(s)
```

**Comment by sellers** [comment]
I had this error when I had the subnetID, subnet CIDR or subnet name not matching exactly what was in AWS.   (e.g. if I said it was a but it was AZ d it was unhappy with this error)   I also had issues (CIDR does not match)  if I had my networkCIDR set to anything except the full CIDR of the entire VPC (even if I was using an  AWS stack or partial section of my VPC)

**Comment by ramonkeypr** [comment]
I am also seeing this issue with:

```
$ kops version
Version 1.7.0 (git-e04c29d)
```

In my case, I added an additional CIDR to my dev vpc on AWS, and now I am attempting to enable some APIs per the instructions in: https://github.com/kubernetes/kops/issues/474

**Comment by chrislovecnm** [comment]
This is a limitation of kops currently.  Changing CIDRs is NOT trivial :(  We are alway happy to have someone help!

**Comment by ramonkeypr** [comment]
Definitely.  But the CIDR for the cluster isn't changing, nor are any of the CIDRs.

**Comment by chrislovecnm** [comment]
Miss understood. What is the vpc definition in the cluster spec?

kops get cluster -o yaml

**Comment by joan38** [comment]
Same here, I don't get was is wrong there

**Comment by dgem** [comment]
Same problem using head:
```
$ kops version
Version 1.8.0-beta.1 (git-8eac35879)
```

**Comment by chrislovecnm** [comment]
Need more details ;)

**Comment by odedpriva** [comment]
I'm trying to create a cluster in an existing VPC that has 6 subnets:
![image](https://user-images.githubusercontent.com/4990615/33566507-da21eb46-d928-11e7-80d5-3f5ab64accdc.png)

I want the new cluster to use its own subnet  172.31.112.0/20
```bash

kops create cluster --name=eu-west-1.k8s.exampleops.com --state=s3://k8s.eu-west-1.exampleops.com --zones=eu-west-1a --vpc=vpc-8c7bfasc --node-size=t2.micro --master-size=t2.micro --dns-zone Z3QL1IADVF67WD --cloud-labels "environment=alphak8s,environment_type=gate1k8s,kops=true" --admin-access $(curl ifconfig.co)/32 --cloud aws --ssh-public-key ~/.ssh/alphak8s.pub --network-cidr 172.31.112.0/20
```
getting the following error:
```bash
I1204 19:21:30.979984   56769 create_cluster.go:971] Using SSH public key: /Users/odedpriva/.ssh/alphak8s.pub
I1204 19:21:32.475987   56769 subnets.go:184] Assigned CIDR 172.31.114.0/23 to subnet eu-west-1a
Previewing changes that will be made:

I1204 19:21:36.885899   56769 executor.go:91] Tasks: 0 done / 71 total; 30 can run
W1204 19:21:37.644847   56769 executor.go:109] error running task "VPC/eu-west-1.k8s.exampleops.com" (9m59s remaining to succeed): Field cannot be changed: CIDR
I1204 19:21:37.644911   56769 executor.go:91] Tasks: 29 done / 71 total; 19 can run
W1204 19:21:38.541841   56769 executor.go:109] error running task "VPC/eu-west-1.k8s.exampleops.com" (9m58s remaining to succeed): Field cannot be changed: CIDR
I1204 19:21:38.541903   56769 executor.go:91] Tasks: 47 done / 71 total; 1 can run
```

```
➜  terraform aws ec2 describe-vpcs --filters "Name=tag:Name,Values=staging" --query 'Vpcs[].CidrBlockAssociationSet[].CidrBlock'
[
    "172.31.0.0/16"
]
➜  terraform kops version
Version 1.8.0 (git-5099bc5)
```

What am I doing wrong ?

**Comment by odedpriva** [comment]
I exported the configuration to yaml, and changed the subnet section. 
honestly, the --network-cidr is a bit confusing.

**Comment by smcquay** [comment]
@chrislovecnm: I'm going to add that this flag has also confused me.

I have an AWS VPC with the CIDR `10.10.0.0/16` and I'd like the kubernetes cluster to only use `10.10.168.0/21` (as much of the network address space is already being used). When I try to run:

```
kops create cluster \
    --cloud=aws \
    --zones us-west-1a \
    --name=k8s-kops-poc-00.infra.poc.aun1.i.wish.com \
    --vpc vpc-<redacted> \
    --networking amazon-vpc-routed-eni \
    --master-size c4.xlarge \
    --kubernetes-version 1.9.2 \
    --network-cidr 10.10.168.0/21 \
    --node-count 4 \
    --node-size m4.2xlarge \
    --state s3://sm-k8s-kops-state-store 
```

I get the same error.

I'm pretty new to the `kops` project (trying it out for the first time) but I'm more than happy to dive into the code and figure out what's not working as we expect and fix it. I'm going to start [here](https://github.com/kubernetes/kops/blob/master/docs/development/adding_a_feature.md), but am happy to engage with whomever/however to design the appropriate solution here.

**Comment by chrislovecnm** [comment]
@smcquay we have office hours tomorrow morning, see the README.  If want to swing by we would love to have you.  Otherwise please reach out via slack if you have questions.  I usually use manifests with more complex configurations, but that is just me.  Can I get a -v10 debugging on the kop create or update?

**Comment by smcquay** [comment]
@chrislovecnm Thanks for pointing it out; I'm planning on being there (in 10 mins).

I'm attaching the files for `create`, which fails, so I was not able to generate the files for `update`

[2830.stderr.txt](https://github.com/kubernetes/kops/files/1732074/2830.stderr.txt)
[2830.stdout.txt](https://github.com/kubernetes/kops/files/1732075/2830.stdout.txt)

**Comment by smcquay** [comment]
I used an old config that doesn't use gossip, I reran with gossip and got the same results:

[2830.stderr.txt](https://github.com/kubernetes/kops/files/1732087/2830.stderr.txt)
[2830.stdout.txt](https://github.com/kubernetes/kops/files/1732088/2830.stdout.txt)

**Comment by fwynyk** [comment]
I'm also trying to update another conf ( `horizontalPodAutoscalerUpscaleDelay`, for instance) and feels to me that `networkCIDR` is not the same as the `CIDR` showed in the error, which is the same problem as @smcquay is talkin about.

I use`--networkCIDR` with a lower range of IPs because I don't want to use all my VPC networks to the cluster. And with `-v-10` I got this:

`I0228 19:00:09.613895   28492 vpc.go:84] found matching VPC *awstasks.VPC {"Name":"xxxxxxxx","Lifecycle":null,"ID":"vpc-xxxxxxx","CIDR":"10.0.0.0/16","EnableDNSHostnames":null,"EnableDNSSupport":null,"Shared":null,"Tags":{}}
I0228 19:00:09.614248   28492 request_logger.go:45] AWS request: ec2/DescribeVpcAttribute
I0228 19:00:09.686204   28492 request_logger.go:45] AWS request: ec2/DescribeVpcAttribute
I0228 19:00:09.736157   28492 changes.go:81] Field changed "CIDR" actual="10.0.0.0/16" expected="10.0.128.0/18"`.

Feels that `10.0.0.0/16` is comming from the vpc configuration (`vpc.go:84` above), not from `networkCIDR` field.

I also asked about it at #kops-users slack channel, with no answer yet.

$ kops version
Version 1.8.0

**Comment by KashifSaadat** [comment]
Hi @odedpriva, in your case (probably also applies to others having similar issues):
- Your VPC CIDR: `172.31.0.0/16`
- Subnets that exist within the above VPC CIDR: (6x /20 blocks looking at your screenshot)
- You are attempting to use a smaller subnet range within the above VPC

There's two different ways to do this:
- Define the existing VPC CIDR + VPC ID in your kops ClusterSpec, and specify what the subnet CIDRs should be (these ranges must reside within the VPC CIDR). kops will attempt to create these subnets, but I THINK it may also do things like changing route tables / creating NAT Gateways, updating tags etc.
- Define the existing VPC CIDR + VPC ID in your kops ClusterSpec, and specify existing AWS Subnet IDs (must be already created). In this scenario, you should make sure that these subnets have all the relevant routing in place (to pull down images, get state from s3 etc).

There's documentation available here on how to do this: https://github.com/kubernetes/kops/blob/master/docs/run_in_existing_vpc.md#shared-subnets

I do something similar and not had any issues so far.

**Comment by bethbuhawe** [comment]
Hi I am having the same issue as what was mentioned in this thread. Basically i can't explicitly specify the CIDR block that i want to use which is already existing in AWS. I can somehow create a cluster if i specify the vpc id, however it seems like it keeps on creating it's own CIDR block such as *.*. 32.0/19. The problem with this CIDR block is that i can't create another cluster on the same region since it keeps using the same 32.0/19. Same thing happens to other AWS regions. Btw, i did try the blob mentioned in this thread but no luck.

**Comment by smcquay** [comment]
@bethbuhawe I had some luck with this. Basically I edit the cluster config before I `kops update cluster`, for example: 

```yaml
spec:
  # other things ...                                                                                                    
  subnets:                                                                                               
  - cidr: 10.0.16.0/20                                                                                   
    name: us-west-1a                                                                                     
    type: Public                                                                                         
    zone: us-west-1a 
```

Having said that it's cumbersome to edit this by hand, so I created a binary that I can use as an `EDITOR` during `kops edit` as:

```bash
$ EDITOR='myprogram' kops edit cluster --name=$KOPS_CLUSTER_NAME
```

My program is essentially this:

```go
package main                                                                                             
                                                                                                         
import (                                                                                                 
    "bytes"                                                                                              
    "io"                                                                                                 
    "fmt"
    "os"                                                                                                 
                                                                                                         
    "github.com/pkg/errors"                                                                              
    api "k8s.io/kops/pkg/apis/kops"                                                                      
    "k8s.io/kops/pkg/kopscodecs"                                                                         
)       

func main() {
  // this is simplified; I'd do validation for os.Args
  if err := edit(os.Args[1]); err != nil {
    fmt.Fprintf(os.Stderr, "edit: %v\n", err)
    os.Exit(1)
  }
}

func edit(fn string) error {                                                                             
    f, err := os.Open(fn)                                                                                
    if err != nil {                                                                                      
        return errors.Wrap(err, "opening input file: %v")                                                
    }                                                                                                    
    buf := &bytes.Buffer{}                                                                               
    if n, err := io.Copy(buf, f); err != nil {                                                           
        return errors.Wrapf(err, "problem reading after %d bytes", n)                                    
    }                                                                                                    
    if err := f.Close(); err != nil {                                                                    
        return errors.Wrap(err, "closing input file")                                                    
    }                                                                                                    
                                                                                                         
    o, _, err := kopscodecs.ParseVersionedYaml(buf.Bytes())                                              
    if err != nil {                                                                                      
        return errors.Wrap(err, "parse yaml")                                                            
    }                                                                                                    
                                                                                                         
    c := o.(*api.Cluster)                                                                                
                                                                                                         
    v, ok := os.LookupEnv(kopsPodSubnet)                                                                 
    if !ok {                                                                                             
        return errors.Wrapf(err, "missing %q variable", kopsPodSubnet)                                   
    }                                                                                                    
                                                                                                         
    if err := validateSubnet(v); err != nil {                                                            
        return errors.Wrap(err, "bad subnet")                                                            
    }                                                                                                    
    
    // Insert your custom modifications here:
                                                                                                     
    c.Spec.Subnets[0].CIDR = v                                                                           
                                                                                                         
    maxPods := kopsMaxPods                                                                               
    kcs := &api.KubeletConfigSpec{                                                                       
        MaxPods: &maxPods,                                                                               
                                                                                                         
        // This is to work around a debian bug in kops.                                                  
        // See: https://github.com/ContextLogic/k8s/issues/3                                             
        KubeletCgroups: kopsCgroups,                                                                     
        RuntimeCgroups: kopsCgroups,                                                                     
                                                                                                         
        CloudProvider: "aws",                                                                            
    }                                                                                                    
                                                                                                         
    c.Spec.Kubelet = kcs                                                                                 
    c.Spec.MasterKubelet = kcs                                                                           
                                                                                                         
    b, err := kopscodecs.ToVersionedYaml(c)                                                              
    if err != nil {                                                                                      
        return errors.Wrap(err, "convert to yaml")                                                       
    }                                                                                                    
    of, err := os.Create(fn)                                                                             
    if err != nil {                                                                                      
        return errors.Wrap(err, "opening output file")                                                   
    }                                                                                                    
    if n, err := of.Write(b); err != nil {                                                               
        return errors.Wrapf(err, "problem writing out yaml after %d bytes:", n)                          
    }                                                                                                    
    return nil                                                                                           
}
```

Hopefully that helps!

**Comment by bethbuhawe** [comment]
Hi @smcquay ,

So i was able to manage this yesterday. The solution i've got so far is more or less the same.

   1st - I created the cluster using kops create cluster but without the --yes option. This enables me to create the files for it's masters and nodes without creating the actual cluster yet.
   2nd - once the cluster files has been created, edit the cluster with kops edit cluster {clustername} and change the cidr blocks to your desired block which is under the cidr block of your own vpc
   3rd - Update your cluster with kops update cluster {clustername} which then creates the actual cluster using AWS resources with your desired cidr block.

Basically, creating and editing kubernetes clusters is quite cumbersome with all the manual things to do. But i'm pretty sure there's another way in dealing with it like the example you've had above. I'm also thinking of managing it if possible via file wherein the configs are ready before the actual creation takes place.

Thanks :)

**Comment by chrislovecnm** [comment]
@bethbuhawe is using manifests and templating not an option for you? See the using manifest documentation. We also have templating in toolbox.

Then you can kops create -f mycluster.yaml

Just a thought

**Comment by smcquay** [comment]
@chrislovecnm It had been suggested to me in slack to use `toolbox` commands and templates, I had some concerns about the format of the templating changing. My assumption was that as I upgrade kops in the future the edit interface would be more stable. Is my reasoning flawed? Is it fine to build up automation around a current snapshot of the templating?

**Comment by chrislovecnm** [comment]
The api is stable, we cannot change it without another API version. The yaml translated directly to the API. When we do a beta release we will have some tweaks, but we are in zero rush for that. I will push to have the alpha v2 api supported for a long long time.

**Comment by chrislovecnm** [comment]
Oh and the edit interface has zero different from using a manifest directly.  We have some validation differences, which we UCD bugs open for, but the API is the API.
